File: rust/day02/src/main.rs
use common::read_lines;

fn is_repeated_pattern(n: u64) -> bool {
    let n_digits = (n as f64).log10() as u64 + 1;

    for pattern_length in 1..=(n_digits/2) {
        if n_digits % pattern_length != 0 {
            continue;
        }

        let divisor = 10u64.pow(pattern_length as u32);
        let pattern = n % divisor;
        
        let mut matches = true;
        let mut temp: u64 = n;

        while temp > 0 {
            if temp % divisor != pattern {
                matches = false;
                break;
            }
            temp /= divisor;
        }
        if matches {
            return true;
        }
    }
    false
}

fn main() {
    let dirty_lines = read_lines("C:\\Users\\marau\\dev\\advent\\Advent2025\\inputs\\day2.txt");
    let mut invalid_value: u64 = 0;
    
    for dirty_line in &dirty_lines {
        for range in dirty_line.trim().split(','){    
            let line = range.trim();
            let low: u64 = line[0..line.find('-').expect("Could not find range")].parse().expect("Could not parse low");
            let high: u64 = line[line.find('-').expect("Could not find range") + 1..].parse().expect("Could not parse high");

            for i in low..=high {
                let n_digits = (i as f64).log10() as u64 + 1;
                if n_digits % 2 == 1 { continue; }
                let first_half = i / 10u64.pow(n_digits as u32 / 2);
                let second_half = i % 10u64.pow(n_digits as u32 / 2);
                if first_half == second_half {
                    invalid_value += i;
                }
            }
        }
    }
    println!("Part 1: {}", invalid_value);

    invalid_value = 0;
    for dirty_line in &dirty_lines {
        for range in dirty_line.trim().split(','){    
            let line = range.trim();
            let low: u64 = line[0..line.find('-').expect("Could not find range")].parse().expect("Could not parse low");
            let high: u64 = line[line.find('-').expect("Could not find range") + 1..].parse().expect("Could not parse high");

            for i in low..=high {
                if is_repeated_pattern(i) {
                    invalid_value += i;
                }
            }
        }
    }
    println!("Part 2: {}", invalid_value);
}

File: go/day02/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func FindRepeatedPatterns(n uint64) bool {
	n_Digits := int(math.Log10(float64(n))) + 1

	for i := 1; i <= n_Digits/2; i++ {
		pattern := n % uint64(math.Pow10(i))

		if n_Digits%i != 0 {
			continue
		}

		repeated := true
		for j := 1; j < n_Digits/i; j++ {
			if n%uint64(math.Pow10((j+1)*i))/uint64(math.Pow10(j*i)) != pattern {
				repeated = false
				break
			}
		}
		if repeated {
			return true
		}
	}
	return false
}

func main() {
	input, err := os.ReadFile("/home/luke/dev/advent/2025/inputs/day2.txt")
	if err != nil {
		panic(err)
	}

	lines := strings.TrimSpace(string(input))

	var invalid uint64 = 0

	for _, line := range strings.Fields(lines) {
		ranges := strings.Split(line, ",")

		for _, r := range ranges {
			parts := strings.Split(r, "-")
			low, _ := strconv.ParseUint(parts[0], 10, 64)
			high, _ := strconv.ParseUint(parts[1], 10, 64)

			for i := low; i <= high; i++ {
				n_Digits := int(math.Log10(float64(i))) + 1
				if n_Digits%2 == 1 {
					continue
				}
				firstHalf := i / uint64(math.Pow10(n_Digits/2))
				secondHalf := i % uint64(math.Pow10(n_Digits/2))
				if firstHalf == secondHalf {
					invalid += i
				}
			}
		}

		fmt.Printf("Part 1: %d\n", invalid)

		invalid = 0
		for _, r := range ranges {
			parts := strings.Split(r, "-")
			low, _ := strconv.ParseUint(parts[0], 10, 64)
			high, _ := strconv.ParseUint(parts[1], 10, 64)

			for i := low; i <= high; i++ {
				if FindRepeatedPatterns(i) {
					invalid += i
				}
			}
		}
		fmt.Printf("Part 2: %d\n", invalid)
	}
}
